### ISSUE TITLE
Creating signals generates no code in external editor

### ISSUE THREAD

**Issue by chucklepie** [godotengine/godot#41283]
***Bugsquad edit:** This issue has been confirmed several times already. No need to confirm it further.*

___

**Godot version:**
3.2.2 stable

**OS/device including version:**
linux mint 64 bit

**Issue description:**
I am using latest VSCode. External editor opens/switches to vscode when accessing code ok. However when I select to create a signal, it sets the signal as created in the godot IDE, it switches to vscode and opens the correct script file, but no code is generated, as per video, it simply jumps to line 1.

Additionally when you double click on the signal to go to the code it doesn't open editor. If I manually create the signal in the correct place, double clicking still does not open the editor. If I go to an existing signal already created then it switches to vscode fine.

As a side note, if I create a new script file and open it, vscode opens and the default templates are populated ok, so it is just signals and nothing to do with creating new code...

![vscode](https://user-images.githubusercontent.com/12863685/90313580-996b5480-df05-11ea-947d-3fd306d27898.gif)

Project settings for vscode are:
code
-g {file}:{line}:{col}

**Steps to reproduce:**
1. create a signal
2. no code is created

**Comment by chucklepie** [comment]
I am now and again (i.e. not always), seeing this message in Godot, despite the files not being open in Godot:

 Another resource is loaded from path 'res://entities/static/Missile_Launcher.gd' (possible cyclic resource inclusion).

**Comment by Calinou** [comment]
This seems to be a duplicate of https://github.com/godotengine/godot/issues/23208, but people are only reporting issues when using C# there.

**Comment by chucklepie** [comment]
That's quite a specific thing to do with the c# class file generation as
well. Maybe this is better separate, but linked by reference?

**Comment by Calinou** [comment]
Cross-posting my comment for reference:

> I think the issue has more to do with using an external editor than C# in itself.

> A simple fix would be to save the file automatically when Godot creates the function in its built-in script editor. This way, the external editor will reload the file (if it supports automatic file reloading) and show the newly created function.

> However, this is problematic if you've made unsaved changes in the external editor, since you'll see a "conflict prompt" in this case (as the file has been modified *externally*, yet you had unsaved changes). This could be alleviated by enabling the "save on focus loss" behavior in your external editor.

**Comment by shawkes** [comment]
> That's quite a specific thing to do with the c# class file generation as well. Maybe this is better separate, but linked by reference?

That aspect (code being generated in the wrong place) is covered in a separate issue https://github.com/godotengine/godot/issues/12908 

This and #23208 are the same underlying issue (PR will fix both), so closing one makes sense.

**Comment by AMOfficalGithub** [comment]
Does anyone know if this issue has been fixed yet?

**Comment by Calinou** [comment]
@AMOfficalGithub Not yet, https://github.com/godotengine/godot/pull/42416 needs to be merged first.

**Comment by samtsejerome** [comment]
This is still an issue in 3.2.3

**Comment by Calinou** [comment]
Reopening because the feature isn't working as intended when an external editor is configured: https://github.com/godotengine/godot/pull/42416#issuecomment-846427354

**Comment by Brin-o** [comment]
I'm on 3.3.2 stable using GDScript + VS Code on Windows and this is an issue as well

**Comment by Pika1998** [comment]
Hi, I am also facing this issue on godot v 3.2.3.stable.mono using C# + VS Code Configured on Windows 10

**Comment by Pika1998** [comment]
Also observed a vice-versa situation where I am declaring a custom-signal in my code from VS Code but unable to see it reflect on the inspector
Code :
![image](https://user-images.githubusercontent.com/36433104/122525089-813cf800-d036-11eb-88a4-a65b81c60995.png)
Inspector :
![image](https://user-images.githubusercontent.com/36433104/122525182-9c0f6c80-d036-11eb-88bf-ac2c939330e5.png)

**Comment by Pika1998** [comment]
Hi @mediad02 , yes it would be really helpful if you attach the screenshots, I will also help to test it out

**Comment by Calinou** [comment]
> I get the same issue when creating a signal, godot does not auto generates the signal code in the external editor.
but if I disconnect the signal, then disable external editor and open the scripts with the internal godot editor and then re-enable external editor and create the signal the code is auto generated within the external editor, VS Code.

This is expected because the built-in script editor will save the script when adding a new signal: https://github.com/godotengine/godot/pull/42416

**Comment by mediad02** [comment]
Sorry for the delayed response, i was taking the screenshots for the response.
i will post the screenshots in gif formal, hopefully the quality is good enough.
![github](https://user-images.githubusercontent.com/20409010/128917143-51d2d2a4-c2aa-4cd9-a4d4-25ec0c14971b.gif)

In conclusion, i think the best solution to this problem at the moment is to find a way for godot to have its external editor enabled and for every script created to be opened in the background by the internal editor and external editor as well.

**Comment by Calinou** [comment]
> In conclusion, i think the best solution to this problem at the moment is to find a way for godot to have its external editor enabled and for every script created to be opened in the background by the internal editor and external editor as well.

This would be very confusing, as you could end up editing the script in the "wrong" editor and not see your changes in the other editor until you save the script. Conflicts will often happen if we do this.

**Comment by mediad02** [comment]
@Calinou it is a tedious process since everytime you create a new script you would have to make sure it is also open within the godot editor, but once its been open and you re-enable external editor there is no need to go back into the internal editor, since opening the scripts from the scenetree or file manager dock always opens them in the external editor.

**Comment by ericlothspeich** [comment]
Still an issue in 3.5

**Comment by otsoaingles** [comment]
I saw this issue today in 3.5.1 when following the "fist 2d game" tutorial.

Godot_v3.5.1-stable_mono_win64
Windows 11
Latest VS Code

**Comment by JohnC3** [comment]
Still an issue in 4.0 and 3.5. Following the first 2d game tutorial. 
Rider
Windows 11 and windows 10.

This issue seems to be quite old it might make sense to document it in the tutorials. Perhaps offer a workaround assuming one works.

**Comment by nrmiller** [comment]
Just saw this is still active. I admit, this confused me since I'm very new to Godot! 😅 

I'm not sure if this helps, but perhaps C# partial classes can be utilized. Instead of modifying the primary script, a Godot-managed partial class might be able to be generated and used.

In general though, it seems like C# grammar parsing may be needed so that Godot can know which class and indentation level to insert the snippet. Or maybe it can search for the `_Ready()` function and insert the event subscription logic too.

If the issue is more so concerned with discarding unsaved changes in the external editor, perhaps we can look into creating a VS Code Extension capable of merging the incoming changes with the unsaved changes.

**Comment by GarrickWinter** [comment]
I also ran into this while following the "Your First 2D Game" tutorial. While the issue itself isn't a devastating problem by any means, I think the fact that it's going to hit newcomers who are following an official tutorial _is_ severe from the perspective of the community's accessibility and growth.

I suggest rewriting the tutorial to explicitly mention the bug and provide a workaround. In particular, without the code generation working, the tutorial's suggestion of leaving the default connection settings seems to clash with the provided C# code. The tutorial suggests connecting to a function called __on_body_entered_, but the C# code presented has a function called _OnBodyEntered_.

It seems after some testing that Godot generates CS files that smooths over the formatting differences between __on_body_entered_ and _OnBodyEntered_, but that's not something I would have expected or understood, and might be something worth contextualizing.

Godot 4.1.1 Mono / Windows 10 / Visual Studio Code

**Comment by Calinou** [comment]
@GarrickWinter Feel free to open a pull request by editing https://github.com/godotengine/godot-docs/blob/master/getting_started/first_2d_game/03.coding_the_player.rst :slightly_smiling_face:

**Comment by GarrickWinter** [comment]
Thanks for prompting! I've made a pull request here. I think a small change to be more explicit, and an explicit warning about the bug, would be sufficient to smooth over this bump in the tutorial flow.

https://github.com/godotengine/godot-docs/pull/7915

**Comment by rickygai** [comment]
Godot version 4.1.1
VS Code version1.82.2
Godot Tools VS Code extension version 1.3.1
System information Windows 11

The problem occurred in Godot 4.1.1 with VSCode as well, even Godot IDE itself is not creating the signal connect routine correctly, the routine is created out of the scope of the class of the script attached such as below.

![269964822-21e4d172-5c58-4206-9fcc-85fa8ef9edf5](https://github.com/godotengine/godot/assets/15626855/b742ceb5-625e-46f9-bfb4-64d2201178a9)

Please help to resolve this problem, it has been there for quite long.

**Comment by Eterxity** [comment]
still does not work as of v4.1.3.stable.mono.official [f06b6836a]

**Comment by Vysp3r** [comment]
This really needs to be added

**Comment by Calinou** [comment]
@Vysp3r Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by EvilBarney** [comment]
Just adding that this is still an issue in Godot 4.2.1 with C# and Visual Studio 2022.

Godot is able to create and open new scripts ok, but connecting a signal doesn't generate any C# code regardless of saving the scene / rebuilding in VS or Godot / closing and re-opening the source file, etc.

**Comment by kastromatos** [comment]
> Edit: as a suggestion, and also given that directly modifying the file may cause problems if it has unsaved changes in the external editor: We already have the ability to copy the signal name to the clipboard from the node panel. How about some way of either copying the full signature (in appropriate language) from the same place, or directly putting the handler code on the clipboard when connecting signals? In the latter case, if it's made obvious enough that that's happening (maybe a "Copy script code to clipboard" checkbox on the create panel), and Godot is still opening the appropriate file, then it's just a case of the user choosing where to hit ctrl-v.

Nice catch but proposed and rejected almost a year ago:
[Crutch for outside editor code generation issue](https://github.com/godotengine/godot-proposals/issues/7314)

**Comment by Calinou** [comment]
> Nice catch but proposed and rejected almost a year ago:

I think this workaround makes a lot of sense, but there's another open proposal tracking it: https://github.com/godotengine/godot-proposals/issues/3633

**Comment by kastromatos** [comment]
@Calinou Yeah... So we have 4 years old annoying bug and 3 years old proposed workaround both not implemented. At least this issue are in top 5 sorted by reaction in editor topic and top 10 in .NET. There is a hope that it will be noticed and fixed soon.

**Comment by Rigggg** [comment]
Just helped someone on discord with this issue. It *seems* like they just didn't have their exec flags configured properly.  This first happened to him in 4.3.  I got him to upgrade to 4.4.1 and the issue still persisted. 

This was their exec flags:

![Image](https://github.com/user-attachments/assets/4ea6a56f-da4e-4f11-8a9a-7f7e9c52aaa1)


This is mine(which works):

![Image](https://github.com/user-attachments/assets/e8ee2cd2-9aaa-4b96-963f-57c45628d944)

After they changed their exec flags, the godot extension for vscode prompted them for the path to their Godot executable.  From what I understand, it hadn't prompted him for the path before.. ??  Seemed odd.

He restarted everything and initially it didn't fix anything. But then he came back to me a minute later saying everything worked.  Since it didn't seem like an immediate fix he said he couldn't point out what exactly fixed it.  

So I guess the takeaway is to at least check your exec flags and make sure you have the godot extension for vscode setup properly?

**Comment by digitalseraphim** [comment]
This bug is about C#, not GodotScript, is that what you're using?

**Comment by Eterxity** [comment]
> Just helped someone on discord with this issue. It _seems_ like they just didn't have their exec flags configured properly. This first happened to him in 4.3. I got him to upgrade to 4.4.1 and the issue still persisted. Windows 10. Only GDScript was mentioned. Nothing to do with C# came up.
> 
> This was their exec flags:
> 
> ![Image](https://github.com/user-attachments/assets/4ea6a56f-da4e-4f11-8a9a-7f7e9c52aaa1)
> 
> This is mine(which works):
> 
> ![Image](https://github.com/user-attachments/assets/e8ee2cd2-9aaa-4b96-963f-57c45628d944)
> 
> After they changed their exec flags, the godot extension for vscode prompted them for the path to their Godot executable. From what I understand, it hadn't prompted him for the path before.
> 
> He restarted everything and initially it didn't fix anything. But then he came back to me a minute later saying everything worked. Since it didn't seem like an immediate fix he said he couldn't point out what exactly fixed it.
> 
> So I guess the takeaway is to at least check your exec flags and make sure you have the godot extension for vscode setup properly?

I've just tried what was done by your friend and I too, got the prompt by VSCode for the path to my godot editor executable. I tried generating a method with both languages and it works with GDScript, but not with C#. This most likely confirms that the issue is with C#, and not the external editor.

**Comment by Rigggg** [comment]
> This bug is about C#, not GodotScript, is that what you're using?
> […](#)
The problem was encountered when using GDScript.  If this is a C# only issue then perhaps [this](https://github.com/godotengine/godot/issues/91101) one needs to be re-opened?  As it was consolidated to this issue.  That was how I found myself here.

**Comment by Eterxity** [comment]
> > This bug is about C#, not GodotScript, is that what you're using?
> > […](#)
> 
> Yes, the problem was encountered when using GDScript. If this is a C# only issue then perhaps [this](https://github.com/godotengine/godot/issues/91101) one needs to be re-opened? As it was consolidated to this issue. That was how I found myself here.

The code generation works fine in VSCode with GDScript when I tried it, so I don't see why #91101 needs to be reopened. No code is generated when using C# however.

**Comment by digitalseraphim** [comment]
I think the point was that it's not related to this ticket, so should be
reopened, separated from this ticket, and the "fix" mentioned earlier in
this thread mentioned on that ticket, and then reclosed closed, so that if
someone comes looking, they find the fix.

**Comment by rickygai** [comment]
This is the reason after I saw this is years problem being ignored, and Godot is an open-source engine, much focus on its GDScript, I decided to use other 3D engine. In fact, I posted similar case to this channel and Godot's forum then ended in being asked to look at its source code myself.

**Comment by NancokPS2** [comment]
If it helps, this also seems to affect Windsurf IDE.  
Edit: The person who was having this problem seems to just have forgotten to give proper permissions to the executable, as they where running it trough steam and didn't give it admin permissions.

**Comment by digitalseraphim** [comment]
> I have this issue with GDScript. Version 4.5.1 on MacOS. I am using Cursor.  It *sometimes* works.  I haven't been able to figure out what makes it stop working.  Although it mostly doesn't work...

That would be an entirely different issue. Code generation for C# is not implemented. Your situation sounds to be an actual bug. You need to either find another issue that more closely matches your situation, or create a new one if you can't find one

**Comment by digitalseraphim** [comment]
So, my bad here, though, partially *someone* else's fault too. I'm following this issue because of the fact that this is not implemented for C#, and new issues about it are getting pointed here. It's a confusing landscape. If you look at https://github.com/godotengine/godot/issues/23208#issuecomment-698117036 they have a workaround which may work for you. 

Also, on rereading the history, I was wrong about it only being a C# issue, it's apparently an "external editor issue". Unfortunately, that workaround will only work for gdscript. Good luck and hope you can get something working

**Comment by DanTulovsky** [comment]
That's totally it... At least this narrows down where the bug is


──────────────────────────────────────
dan-generated content may be incorrect
